ggu/main.c

```diff
@@ -246,83 +373,32 @@ process_single_packet(struct rte_mbuf *pkt, const struct ggu_config *ggu_conf)
 		goto free_packet;
 	}
 
-	/* XXX Check the UDP checksum. */
-
-	gguhdr = (struct ggu_common_hdr *)&udphdr[1];
-	if (gguhdr->v1 != GGU_PD_VER1) {
-		RTE_LOG(NOTICE, GATEKEEPER,
-			"ggu: unknown policy decision format %hhu\n",
-			gguhdr->v1);
-		goto free_packet;
-	}
-
-	policy_ptr = (uint8_t *)&gguhdr[1];
-
-	real_payload_len = rte_be_to_cpu_16(udphdr->dgram_len);
-	expected_payload_len = sizeof(*udphdr) + sizeof(*gguhdr) +
-		(gguhdr->n1 + gguhdr->n3) * sizeof(policy.flow.f.v4) +
-		(gguhdr->n2 + gguhdr->n4) * sizeof(policy.flow.f.v6) +
-		(gguhdr->n1 + gguhdr->n2) * sizeof(policy.params.u.declined) + 
-		(gguhdr->n3 + gguhdr->n4) * sizeof(policy.params.u.granted);
-	if (real_payload_len < expected_payload_len) {
+	real_payload_len = pkt->data_len - l2_len - l3_len;
+	expected_payload_len = rte_be_to_cpu_16(udphdr->dgram_len);
+	if (real_payload_len != expected_payload_len) {
 		RTE_LOG(NOTICE, GATEKEEPER,
 			"ggu: the size (%hu) of the payload available in the UDP header doesn't match the expected size (%hu)!\n",
 			real_payload_len, expected_payload_len);
 		goto free_packet;
 	}
 
-	/* Loop over each policy decision on the packet. */
-
-	/* Process the IPv4 decline decisions. */
-	memset(&policy, 0, sizeof(policy));
-	policy.state = GK_DECLINED;
-	policy.flow.proto = ETHER_TYPE_IPv4;
-	for (j = 0; j < gguhdr->n1; j++) {
-		rte_memcpy(&policy.flow.f.v4, policy_ptr,
-			sizeof(policy.flow.f.v4));
-		policy_ptr += sizeof(policy.flow.f.v4);
-		rte_memcpy(&policy.params.u.declined, policy_ptr,
-			sizeof(policy.params.u.declined));
-		policy_ptr += sizeof(policy.params.u.declined);
-		process_single_policy(&policy, ggu_conf);
-	}
+	/* XXX Check the UDP checksum. */
 
-	/* Process the IPv6 decline decisions. */
-	policy.flow.proto = ETHER_TYPE_IPv6;
-	for (j = 0; j < gguhdr->n2; j++) {
-		rte_memcpy(&policy.flow.f.v6, policy_ptr,
-			sizeof(policy.flow.f.v6));
-		policy_ptr += sizeof(policy.flow.f.v6);
-		rte_memcpy(&policy.params.u.declined, policy_ptr,
-			sizeof(policy.params.u.declined));
-		policy_ptr += sizeof(policy.params.u.declined);
-		process_single_policy(&policy, ggu_conf);
+	gguhdr = (struct ggu_common_hdr *)&udphdr[1];
+	if (gguhdr->version != GGU_PD_VER) {
```

Comment:
Let's check that the reserved fields are zero as well. It'll help to rewrite `uint8_t reserved[3];` as `uint8_t res1; uint16_t res2;` as you did in `struct ggu_decision`.

Reply:
You forgot to write the verification.